sweepbatcher/presigned.go

```diff
@@ -0,0 +1,602 @@
+	const feeRate = chainfee.FeePerKwFloor
+
+	tx, _, _, _, err := constructUnsignedTx(
+		sweeps, destAddr, currentHeight, feeRate,
+	)
+	if err != nil {
+		return fmt.Errorf("failed to construct unsigned tx "+
+			"for feeRate %v: %w", feeRate, err)
+	}
+
+	// Check of a presigned transaction exists.
+	var batchAmt btcutil.Amount
+	for _, sweep := range newSweeps {
+		batchAmt += sweep.value
+	}
+	const loadOnly = true
+	signedTx, err := b.cfg.presignedHelper.SignTx(
+		ctx, primarySweepID, tx, batchAmt, feeRate, feeRate, loadOnly,
+	)
+	if err != nil {
+		return fmt.Errorf("failed to find a presigned transaction "+
+			"for feeRate %v, txid of the template is %v, inputs: %d, "+
+			"outputs: %d: %w", feeRate, tx.TxHash(),
+			len(tx.TxIn), len(tx.TxOut), err)
+	}
+
+	// Check the SignTx worked correctly.
+	err = CheckSignedTx(tx, signedTx, batchAmt, feeRate)
+	if err != nil {
+		return fmt.Errorf("signed tx doesn't correspond the "+
+			"unsigned tx: %w", err)
+	}
+
+	return nil
+}
+
+// getOrderedSweeps returns the sweeps of the batch in the order they were
+// added. The method must be called from the event loop of the batch.
+func (b *batch) getOrderedSweeps(ctx context.Context) ([]sweep, error) {
+	// We use the DB just to know the order. Sweeps are copied from RAM.
+	utxo2sweep := make(map[wire.OutPoint]sweep, len(b.sweeps))
+	for _, s := range b.sweeps {
+		utxo2sweep[s.outpoint] = s
+	}
+
+	dbSweeps, err := b.store.FetchBatchSweeps(ctx, b.id)
+	if err != nil {
+		return nil, fmt.Errorf("FetchBatchSweeps(%d) failed: %w", b.id,
+			err)
+	}
+	if len(dbSweeps) != len(utxo2sweep) {
+		return nil, fmt.Errorf("FetchBatchSweeps(%d) returned %d "+
+			"sweeps, len(b.sweeps) is %d", b.id, len(dbSweeps),
+			len(utxo2sweep))
+	}
+
+	orderedSweeps := make([]sweep, len(dbSweeps))
+	for i, dbSweep := range dbSweeps {
+		// Sanity check: make sure dbSweep.ID grows.
+		if i > 0 && dbSweep.ID <= dbSweeps[i-1].ID {
+			return nil, fmt.Errorf("sweep ID does not grow: %d->%d",
+				dbSweeps[i-1].ID, dbSweep.ID)
+		}
+
+		s, has := utxo2sweep[dbSweep.Outpoint]
+		if !has {
+			return nil, fmt.Errorf("FetchBatchSweeps(%d) returned "+
+				"unknown sweep %v", b.id, dbSweep.Outpoint)
+		}
+		orderedSweeps[i] = s
+	}
+
+	return orderedSweeps, nil
+}
+
+// getSweepsGroups returns groups in which sweeps were added to the batch.
+// All the sweeps are sorted by addition order and groupped by swap.
+// The method must be called from the event loop of the batch.
+func (b *batch) getSweepsGroups(ctx context.Context) ([][]sweep, error) {
+	orderedSweeps, err := b.getOrderedSweeps(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("getOrderedSweeps(%d) failed: %w", b.id,
+			err)
+	}
+
+	groups := [][]sweep{}
+	for _, s := range orderedSweeps {
+		// Start new group if there are no groups or new swap starts.
+		if len(groups) == 0 ||
+			s.swapHash != groups[len(groups)-1][0].swapHash {
+
+			groups = append(groups, []sweep{})
+		}
+
+		index := len(groups) - 1
```

Comment:
move this to the start of the for loop and use it in the `if`.

Reply:
Fixed.

```go
groups := [][]sweep{}
for _, s := range orderedSweeps {
		index := len(groups) - 1

		// Start new group if there are no groups or new swap starts.
		if len(groups) == 0 || s.swapHash != groups[index][0].swapHash {
				groups = append(groups, []sweep{})
				index++
		}

		groups[index] = append(groups[index], s)
}
```